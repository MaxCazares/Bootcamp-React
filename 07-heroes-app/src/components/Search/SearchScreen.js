import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import useForm from '../../hooks/useForm';
import getHeroesByName from '../../selectors/getHeroesByName';
import HeroCard from '../Heroes/HeroCard';
import { useMemo } from 'react';

const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const {q = ''} = queryString.parse(location.search);

    const [values, handleInputchange] = useForm({
        searchText: q
    });

    const {searchText} = values;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = e => {
        e.preventDefault();
        navigate(`?q=${searchText}`);
    };

    return (
        <div className='container mt-5'>
            <h1>Búsquedas</h1>
            <hr/>
            <div className='row'> 
                <div className='col-5'>
                    <h4>Buscar</h4>
                    <hr/>
                    <form className='d-grid gap-2' onSubmit={handleSearch}>
                        <input 
                            type='text' 
                            placeholder='Buscar un héroe'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={searchText}
                            onChange={handleInputchange}
                        />
                        <button 
                            type='submit' 
                            className='btn btn-outline-warning mt-1'>
                            Buscar
                        </button>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Resultados</h4>
                    <hr/>
                    {
                        (q === '') 
                            ? <div className='alert alert-info'>Buscar un héroe</div>
                            : (heroesFiltered.length === 0) 
                                && <div className='alert alert-danger'>No hay resultados</div>
                    }
                    {
                        heroesFiltered.map(hero => (
                            <HeroCard key={hero.id} {...hero}/>
                        ))
                    }
                </div>
            </div>
        </div>
    );
}
 
export default SearchScreen;