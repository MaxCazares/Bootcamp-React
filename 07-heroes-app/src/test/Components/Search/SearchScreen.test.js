import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import SearchScreen from '../../../components/Search/SearchScreen';

describe('Pruebas en <SearchScreen/>', () => {
    
    test('Debe mostrarse correctamente', () => {
        
        const wrapper = mount(
            <MemoryRouter>
                <SearchScreen/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un hÃ©roe');

    });

    test('Debe mostrar a Batman y el input con el valor de queryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?=Batman']}>
                <SearchScreen/>
            </MemoryRouter>
        );

        // expect(wrapper.find('input').prop('value')).toBe('Batman');
        console.log(wrapper.find('input').prop('value'));

    });

});