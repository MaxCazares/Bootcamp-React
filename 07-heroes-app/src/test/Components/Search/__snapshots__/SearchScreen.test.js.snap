// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe mostrarse correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Búsquedas
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        className="d-grid gap-2"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-warning mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>
</div>
`;
